refactor: cleanup compiler pipeline

bartlomieju opened the pull request:
Closes #2657 and #2658

Depends on #2683

bartlomieju commented:
Okay I addressed all the things I wanted to in this PR. 

Brief summary:
- remove `fetch_source_file_and_maybe_compile_async` and replace it with `State.fetch_compiled_module`
- introduce `CompiledModule` which is basically the same as `deno::SourceInfo` and represents arbitrary file that has been compiled to JS module
- introduce `//cli/compilers` module containing all compilers
- introduce `JsCompiler`, `JsonCompiler` encapsulating transforming `SourceFile` into `CompiledModule`

Implemented solution allows to handle other media types rather easily:
1. Add another media type to `msg::MediaType`
2. Create compiler structure implementing `compile_async` method that takes raw source file and produces compiled JS module out of it
3. Wire the compiler in `State.fetch_compiled_module`

This PR is still missing tests, but I first want to gather feedback regarding this refactor before doing that.

CC @ry @piscisaureus @kitsonk @kevinkassimo

ry commented on cli/compiler.rs:
```diff
@@ -27,10 +28,78 @@ use std::sync::atomic::Ordering;
+
+impl CompilerConfig {
+  pub fn load(config_path: Option<String>) -> Result<Self, ErrBox> {
+    // take the passed flag and resolve the file name relative to the cwd
```
Seems like this should be a triple slash comment

ry commented on cli/compiler.rs:
```diff
@@ -121,81 +190,50 @@ pub fn source_code_version_hash(
+    // If `checkJs` is set to true in `compilerOptions` then we're gonna be compiling
+    // JavaScript files as well
+    let config_json = config.json()?;
+    let compile_js = match &config_json.get("compilerOptions") {
```
Is the config parsed elsewhere in rust land ? If so maybe it makes sense to cache the parsed json object?

bartlomieju replied:
No, that's the only place in Rust

kitsonk replied:
Just a health warning, `tsconfig.json` supports JSONC, which serde does not support.  This means that it will not be able to parse certain _valid_ `tsconfig.json` files.  When you scaffold a `tsconfig.json` via the `tsc` it includes comments, and so I think it will be a pretty common pattern in the wild to use JSONC.  So detecting `checkJs` may need to be done in another way, likely a regular expression of some sort.  It also might make sense to have the test fixtures have a `tsconfig.json` that includes comments.

ry commented on cli/compilers/js.rs:
```diff
@@ -0,0 +1,24 @@
+
+pub struct JsCompiler {}
+
+impl JsCompiler {
```
The idea is that in the future these will be trait impl ?

bartlomieju replied:
Yes, exactly. I'll try to refactor TsCompiler not to require `state` arg during compilation and factor out `Compiler` trait

bartlomieju replied:
This gonna need some more thought... We're updating metrics on `state` when spinning up TS compiler so ATM it looks like `state` is a required argument. I'm gonna leave it as is now and addresses in follow up PR.

bartlomieju commented:
Ready for review @ry

ry commented on cli/compilers/js.rs:
```diff
@@ -0,0 +1,24 @@
+pub struct JsCompiler {}
+
+impl JsCompiler {
+  pub fn compile_async(
```
All of these `compile_async` methods don't seem to do anything async. It's passed a `source_file`, which already contains the data. It seems to just wrap it up in a FutureResult... is there a more appropriate name for this method?

ry commented on cli/compilers/ts.rs:
```diff
@@ -290,22 +329,12 @@ impl TsCompiler {
     self: &Self,
```
Oops - I see that the real "compile_async" is here.. ok ignore my previous comment.

ry commented:
Is there a test showing `support JS-to-JS compilation using checkJs`? It doesn't seem so..

ry commented on cli/compilers/mod.rs:
```diff
@@ -0,0 +1,15 @@
+
+pub mod js;
+pub mod json;
+pub mod ts;
```
Consider doing this instead:
```
pub use js::JsCompiler;
pub use json::JsonCompiler;
pub use ts::TsCompiler;
```
Then we don't expose these submodules to the caller

ry commented on cli/compilers/json.rs:
```diff
@@ -0,0 +1,25 @@
+use crate::compilers::CompiledModule;
```
add copyright headers to this file and the others in cli/compiles
```
// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
```

bartlomieju commented:
> Is there a test showing `support JS-to-JS compilation using checkJs`? It doesn't seem so..

@ry addressed feedback and added `tests/038_checkjs.test` which ensures that JS files are run through TS compiler when `checkJs` is on

ry commented on tests/038_checkjs.js:
```diff
@@ -0,0 +1,6 @@
+// console.log intentionally misspelled to trigger a type error
+consol.log("hello world!");
```
👍 Thanks
